pkg/contracts/sui/gateway.go

```diff
@@ -57,22 +67,61 @@ func ActiveMessageContextDynamicFieldName() (json.RawMessage, error) {
 	return dynamicFieldNameToJSONArray("active_message_context")
 }
 
-// NewGatewayFromPairID creates a new Sui Gateway
-// from pair of `$packageID,$gatewayObjectID`
-func NewGatewayFromPairID(pair string) (*Gateway, error) {
-	packageID, gatewayObjectID, err := parsePair(pair)
+// NewGatewayFromAddress creates a new Sui Gateway struct from the gateway address in chain params.
+// The gateway address is in the form of `$packageID,$gatewayObjectID[,$withdrawCapID,originalPackageID]`
+func NewGatewayFromAddress(gatewayAddress string) (*Gateway, error) {
+	packageID, gatewayObjectID, withdrawCapID, originalPackageID, err := parseGatewayAddress(gatewayAddress)
 	if err != nil {
 		return nil, err
 	}
 
-	return NewGateway(packageID, gatewayObjectID), nil
+	return &Gateway{
+		packageID:         packageID,
+		objectID:          gatewayObjectID,
+		withdrawCapID:     withdrawCapID,
+		originalPackageID: originalPackageID,
```

Comment:
Should not be named `originalPackageID` but `previousPackageID`

We keep track of n-1, not the originall

Reply:
Not sure how correct will be the prefix `previous`. For say, we upgrade the package in future again at some point (e.g. V3 added more methods), the `previousPackageID` should change and points to V2?

Reply:
renamed in same commit https://github.com/zeta-chain/node/pull/4138/commits/a49f613fce8b5775b0279ff52b79cedfed533980